sequencer/statetransition.go

```diff
@@ -0,0 +1,161 @@
+package sequencer
+
+import (
+	"fmt"
+	"maps"
+	"time"
+
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/std/algebra/emulated/sw_bw6761"
+
+	stdgroth16 "github.com/consensys/gnark/std/recursion/groth16"
+	"github.com/vocdoni/arbo/memdb"
+	"github.com/vocdoni/vocdoni-z-sandbox/circuits"
+	"github.com/vocdoni/vocdoni-z-sandbox/circuits/statetransition"
+	"github.com/vocdoni/vocdoni-z-sandbox/log"
+	"github.com/vocdoni/vocdoni-z-sandbox/state"
+	"github.com/vocdoni/vocdoni-z-sandbox/storage"
+	"github.com/vocdoni/vocdoni-z-sandbox/types"
+)
+
+func (s *Sequencer) startStateTransitionProcessor() error {
+	const tickInterval = time.Second
+	ticker := time.NewTicker(tickInterval)
+
+	go func() {
+		defer ticker.Stop()
+		log.Infow("state transition processor started",
+			"tickInterval", tickInterval)
+
+		for {
+			select {
+			case <-s.ctx.Done():
+				log.Infow("state transition processor stopped")
+				return
+			case <-ticker.C:
+				s.processPendingTransitions()
+			}
+		}
+	}()
+	return nil
+}
+
+func (s *Sequencer) processPendingTransitions() {
+	// Copy pids to avoid locking the map for too long
+	s.pidsLock.RLock()
+	pids := make(map[string]time.Time, len(s.pids))
+	maps.Copy(pids, s.pids)
+	s.pidsLock.RUnlock()
+
+	// iterate over the process IDs and process the ones that are ready
+	for pid := range pids {
+		// check if there is a batch ready for processing
+		batch, _, err := s.stg.NextBallotBatch([]byte(pid))
+		if err != nil {
+			if err != storage.ErrNoMoreElements {
+				log.Errorw(err, "failed to get next ballot batch")
+			}
+			continue
+		}
+		// if the batch is nil, skip it
+		if batch == nil {
+			continue
+		}
+		// decode process ID and load metadata
+		processID := new(types.ProcessID).SetBytes(batch.ProcessID)
+		log.Debugw("state transition ready for processing",
+			"processID", processID.String(),
+			"ballotCount", len(batch.Ballots))
+		startTime := time.Now()
+		// process the batch to get the proof
+		// proof, err := s.processStateTransitionBatch(processID, batch)
+		_, err = s.processStateTransitionBatch(processID, batch)
+		if err != nil {
+			log.Errorw(err, "failed to process state transition batch")
+			continue
+		}
+		log.Debugw("state transition proof generated", "took", time.Since(startTime).String())
+
+		// TODO: publish the new state to the smart contract with the proof
+
+		// Update the last update time
+		s.pidsLock.Lock()
+		s.pids[pid] = time.Now()
+		s.pidsLock.Unlock()
+	}
+}
+
+func (s *Sequencer) processStateTransitionBatch(
+	processID *types.ProcessID,
+	batch *storage.AggregatorBallotBatch,
+) (groth16.Proof, error) {
+	// initialize the process state
+	processState, err := s.startStateFromSmartContract(processID)
```

Comment:
I think the state-transition processor should not fetch the state from the smart contract. I'd consider that the current local state is always up-to-date. Then create another background processor that keeps updated the local state, in case there is a change on the smart contract that we don't have (this cannot happen on the current MVP, because we are dealing with a single sequencer, but it will be prepared for the future).

So, once the new state-transition is created, it would be queued (same way as the vote batches).